clippy_lints/src/static_items_large_align.rs

```diff
@@ -0,0 +1,92 @@
+use clippy_utils::diagnostics::span_lint_and_note;
+use rustc_hir::{Item, ItemKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{self, AdtDef, Ty, TyCtxt};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_target::abi::Align;
+use rustc_typeck::hir_ty_to_ty;
+
+pub struct StaticItemsLargeAlign {
+    pub page_size: u64,
+}
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Check for usage of static items which have type alignment larger than page size.
+    ///
+    /// ### Why is this bad?
+    /// Due to some known unsound issues, the type alignment may not be fulfilled.
+    /// For more information, see:
+    /// <https://github.com/rust-lang/rust/issues/70022> and
+    /// <https://github.com/rust-lang/rust/issues/70143>.
+    ///
+    /// ### Example
+    /// ```rust
+    /// #[repr(align(0x100000))]
+    /// struct Aligned(u8);
+    ///
+    /// static X: Aligned = Aligned(0);         // Bad
+    ///
+    /// fn main() {
+    ///     let x = Aligned(0);                 // Good
+    ///     println!("{:#x}", &x as *const _ as usize);
+    ///     println!("{:#x}", &X as *const _ as usize);
+    ///     let b = Box::new(Aligned(0));       // Good
+    ///     println!("{:#x}", Box::into_raw(b) as usize);
+    /// }
+    #[clippy::version = "1.61.0"]
+    pub STATIC_ITEMS_LARGE_ALIGN,
+    pedantic,
+    "static items with large type alignment, which may trigger unsound problems"
+}
+impl_lint_pass!(StaticItemsLargeAlign => [STATIC_ITEMS_LARGE_ALIGN]);
+
+impl LateLintPass<'_> for StaticItemsLargeAlign {
+    fn check_item(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        if_chain! {
+            if let ItemKind::Static(hir_ty, _, _) = item.kind;
+            let ty = hir_ty_to_ty(cx.tcx, hir_ty);
+            if let Some(adt_ref) = self.check_ty_alignment(cx.tcx, ty);
+            then {
+                span_lint_and_note(
+                    cx,
+                    STATIC_ITEMS_LARGE_ALIGN,
+                    item.span,
+                    "this static item (itself or its subfield) has large type alignment, which may not be fulfilled,\n\
+                    for more information, see <https://github.com/rust-lang/rust/issues/70022>",
+                    Some(cx.tcx.def_span(adt_ref.did())),
+                    format!("this type has an alignment larger than page size ({}KB)", self.page_size/1024).as_str()
+                );
+            }
+        }
+    }
+}
+
+impl StaticItemsLargeAlign {
+    /// It checks a type with the following steps:
+    /// 1. picks out this type if its kind is among adt, array, tuple or ref to them;
+    ///    otherwise return None
+    /// 2. check if its (or its inner fields') alignment are larger than page size
+    /// 3. return one of them;
+    ///    otherwise return None
+    fn check_ty_alignment<'tcx>(&self, tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> Option<AdtDef<'tcx>> {
```

Comment:
I wonder if we can do a bit better on the spans and also collect the spans of all fields of intermediate structs. Not sure how to get a struct field's span though.

Reply:
How about just report the intermediate types in mir but without the spans (from hir)? The difficult part in collecting the spans is dealing with the generics in hir.

A simple example:
```rust
#[repr(align(0x100000))]
struct Aligned(u8);

struct AlignedWrapper {
    f: u8,
    g: Aligned,
}

struct AlignedGeneric<T>(T);

static XG: AlignedGeneric<AlignedWrapper> = ...;
```
Then the intermediate types in mir would be:
```rust
AlignedGeneric<AlignedWrapper>
AlignedWrapper
Aligned
```
But in hir we can only get:
```rust
AlignedGeneric<T>(T)
// T needs to be resolved to AlignedWrapper
// and it is not straightforward to do it in hir
...
```

As we get more involved in connecting the mir and hir (the solution I have been trying), it is likely the lint/checking would become unmanageable and be unsound (i.e., having false negative).

Here is a more involved example (as a reference):
```rust
#[repr(align(0x100000))]
struct AlignedA(u8);

#[repr(align(0x100000))]
struct AlignedB(u8);

struct FnPtr<T>(fn() -> Box<T>);

struct AG<T>(T);

type AGAlias<T> = AG<T>;

struct AGWithArgs<A, B> {
    not_aligned: FnPtr<A>,
    aligned: AGAlias<B>,
}

static XG: AGWithArgs<AlignedA, AlignedB> = AGWithArgs {
    not_aligned: FnPtr(box_aligned),
    aligned: AG(AlignedB(0)),
};

fn box_aligned() -> Box<AlignedA> {
    Box::new(AlignedA(0))
}
```
The report in mir would be straightforward (since it is more easy to mono mir types):
```rust
AGWithArgs<AlignedA, AlignedB>
AG(AlignedB(0))
AlignedB
```
But it is difficult to report the relevant spans (from hir) like this:
```rust
struct AGWithArgs<A, B> {
    not_aligned: FnPtr<A>,
    aligned: AGAlias<B>,
    ^^^^^^^^^^^^^^^^^^^
}
type AGAlias<T> = AG<T>;
                  ^^^^^
struct AG<T>(T);
            ^^^
AlignedB
```

Or maybe there is a good way to work with (e.g., some subst mechanisms) generics in hir I am missing? And now I also start to wonder how rustc deal with generics when reporting (type) errors to user.

Reply:
Yea, that seems reasonable. You can use `tcx.def_span` to get the span of the structs.